File: pedometer.py
# pedometer.py
import json
import os
import argparse
from datetime import datetime, date

CONFIG_FILE = "pedometer_config.json"
DATA_FILE = "steps.json"

class Pedometer:
    def __init__(self):
        self.stride_cm = 78.0
        self.weight_kg = 70.0
        self.goal = 10000
        self.entries = []
        self.load_config()
        self.load_entries()

    def load_config(self):
        if os.path.exists(CONFIG_FILE):
            with open(CONFIG_FILE, "r") as f:
                cfg = json.load(f)
                self.stride_cm = cfg.get("stride_cm", 78.0)
                self.weight_kg = cfg.get("weight_kg", 70.0)
                self.goal = cfg.get("goal", 10000)

    def save_config(self):
        with open(CONFIG_FILE, "w") as f:
            json.dump({
                "stride_cm": self.stride_cm,
                "weight_kg": self.weight_kg,
                "goal": self.goal
            }, f)

    def load_entries(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                self.entries = json.load(f)

    def save_entries(self):
        with open(DATA_FILE, "w") as f:
            json.dump(self.entries, f, indent=2)

    def get_entry(self, date_str):
        for e in self.entries:
            if e["date"] == date_str:
                return e
        return None

    def add_steps(self, steps, date_str=None):
        if date_str is None:
            date_str = datetime.now().strftime("%Y-%m-%d")
        entry = self.get_entry(date_str)
        if entry:
            entry["steps"] = steps
        else:
            self.entries.append({"date": date_str, "steps": steps})
        self.save_entries()
        dist = (steps * self.stride_cm) / 100000.0  # km
        cal = steps * 0.04 * self.weight_kg / 70.0  # rough estimate
        print(f"✅ Logged {steps:,} steps on {date_str} – {dist:.2f} km, {cal:.0f} kcal")

    def stats(self):
        if not self.entries:
            print("No entries.")
            return
        total_steps = sum(e["steps"] for e in self.entries)
        total_dist = (total_steps * self.stride_cm) / 100000.0
        total_cal = total_steps * 0.04 * self.weight_kg / 70.0
        avg_steps = total_steps // len(self.entries)
        today = datetime.now().strftime("%Y-%m-%d")
        today_entry = self.get_entry(today)
        today_steps = today_entry["steps"] if today_entry else 0
        pct = min(100, int(today_steps / self.goal * 100))

        # Energy saving: walking vs driving (approx: 1 L fuel per 10 km)
        fuel_saved = total_dist / 10.0
        co2_saved = fuel_saved * 2.31  # kg CO2 per liter

        print(f"\n🚶 Pedometer (Energy Saving)")
        print(f"Today: {today_steps:,} steps")
        print(f"Goal: {self.goal:,} steps ({pct}% complete)")
        print(f"\n📊 Statistics")
        print(f"Total steps: {total_steps:,}")
        print(f"Total distance: {total_dist:.2f} km")
        print(f"Total calories: {total_cal:.0f} kcal")
        print(f"Average: {avg_steps:,} steps/day")
        print(f"Energy saved: {fuel_saved:.2f} L of fuel")
        print(f"CO₂ saved: {co2_saved:.2f} kg")

    def list(self):
        if not self.entries:
            print("No entries.")
            return
        print("\n📋 Daily Log:")
        for e in sorted(self.entries, key=lambda x: x["date"], reverse=True):
            steps = e["steps"]
            dist = (steps * self.stride_cm) / 100000.0
            cal = steps * 0.04 * self.weight_kg / 70.0
            pct = min(100, int(steps / self.goal * 100))
            print(f"{e['date']} | {steps:,} steps | {dist:.2f} km | {cal:.0f} kcal | {pct}%")

    def export(self, filename):
        import csv
        with open(filename, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(["Date", "Steps", "Distance (km)", "Calories", "Goal %"])
            for e in self.entries:
                steps = e["steps"]
                dist = (steps * self.stride_cm) / 100000.0
                cal = steps * 0.04 * self.weight_kg / 70.0
                pct = min(100, int(steps / self.goal * 100))
                writer.writerow([e["date"], steps, round(dist, 2), round(cal, 0), pct])
        print(f"✅ Exported {len(self.entries)} entries to {filename}")

def main():
    parser = argparse.ArgumentParser(description="Pedometer (Energy Saving)")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    config_parser = subparsers.add_parser("config")
    config_parser.add_argument("--stride", type=float, help="Stride length in cm")
    config_parser.add_argument("--weight", type=float, help="Weight in kg")
    config_parser.add_argument("--goal", type=int, help="Daily step goal")

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("steps", type=int)
    add_parser.add_argument("--date", help="YYYY-MM-DD")

    subparsers.add_parser("stats")
    subparsers.add_parser("list")

    export_parser = subparsers.add_parser("export")
    export_parser.add_argument("filename", default="steps.csv", nargs="?")

    args = parser.parse_args()
    ped = Pedometer()

    if args.cmd == "config":
        if args.stride:
            ped.stride_cm = args.stride
        if args.weight:
            ped.weight_kg = args.weight
        if args.goal:
            ped.goal = args.goal
        ped.save_config()
        print(f"✅ Config updated: stride={ped.stride_cm}cm, weight={ped.weight_kg}kg, goal={ped.goal}")
    elif args.cmd == "add":
        ped.add_steps(args.steps, args.date)
    elif args.cmd == "stats":
        ped.stats()
    elif args.cmd == "list":
        ped.list()
    elif args.cmd == "export":
        ped.export(args.filename)

if __name__ == "__main__":
    main()

File: Pedometer.cs
// Pedometer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Config
{
    [JsonPropertyName("stride_cm")]
    public double StrideCm { get; set; } = 78.0;
    [JsonPropertyName("weight_kg")]
    public double WeightKg { get; set; } = 70.0;
    [JsonPropertyName("goal")]
    public int Goal { get; set; } = 10000;
}

class Entry
{
    [JsonPropertyName("date")]
    public string Date { get; set; }
    [JsonPropertyName("steps")]
    public int Steps { get; set; }
}

class Pedometer
{
    private static readonly string ConfigFile = "pedometer_config.json";
    private static readonly string DataFile = "steps.json";
    private static readonly JsonSerializerOptions Options = new JsonSerializerOptions { WriteIndented = true };

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: Pedometer [config|add|stats|list|export]");
            return;
        }
        Config cfg = LoadConfig();
        List<Entry> entries = LoadEntries();
        string cmd = args[0];

        switch (cmd)
        {
            case "config":
                for (int i = 1; i < args.Length; i++)
                {
                    if (args[i] == "--stride" && i+1 < args.Length)
                        cfg.StrideCm = double.Parse(args[++i]);
                    if (args[i] == "--weight" && i+1 < args.Length)
                        cfg.WeightKg = double.Parse(args[++i]);
                    if (args[i] == "--goal" && i+1 < args.Length)
                        cfg.Goal = int.Parse(args[++i]);
                }
                SaveConfig(cfg);
                Console.WriteLine($"✅ Config updated: stride={cfg.StrideCm}cm, weight={cfg.WeightKg}kg, goal={cfg.Goal}");
                break;

            case "add":
                if (args.Length < 2) { Console.WriteLine("Usage: add <steps> [--date YYYY-MM-DD]"); return; }
                int steps = int.Parse(args[1]);
                string dateStr = null;
                for (int i = 2; i < args.Length; i++)
                {
                    if (args[i] == "--date" && i+1 < args.Length)
                        dateStr = args[++i];
                }
                entries = AddSteps(entries, cfg, steps, dateStr);
                break;

            case "stats":
                Stats(entries, cfg);
                break;

            case "list":
                ListEntries(entries, cfg);
                break;

            case "export":
                string filename = args.Length > 1 ? args[1] : "steps.csv";
                ExportCSV(entries, cfg, filename);
                break;

            default:
                Console.WriteLine("Unknown command");
                break;
        }
    }

    static Config LoadConfig()
    {
        if (!File.Exists(ConfigFile)) return new Config();
        string json = File.ReadAllText(ConfigFile);
        return JsonSerializer.Deserialize<Config>(json) ?? new Config();
    }

    static void SaveConfig(Config cfg)
    {
        string json = JsonSerializer.Serialize(cfg, Options);
        File.WriteAllText(ConfigFile, json);
    }

    static List<Entry> LoadEntries()
    {
        if (!File.Exists(DataFile)) return new List<Entry>();
        string json = File.ReadAllText(DataFile);
        return JsonSerializer.Deserialize<List<Entry>>(json) ?? new List<Entry>();
    }

    static void SaveEntries(List<Entry> entries)
    {
        string json = JsonSerializer.Serialize(entries, Options);
        File.WriteAllText(DataFile, json);
    }

    static Entry GetEntry(List<Entry> entries, string date)
    {
        return entries.FirstOrDefault(e => e.Date == date);
    }

    static List<Entry> AddSteps(List<Entry> entries, Config cfg, int steps, string dateStr)
    {
        if (string.IsNullOrEmpty(dateStr)) dateStr = DateTime.Now.ToString("yyyy-MM-dd");
        var entry = GetEntry(entries, dateStr);
        if (entry != null)
        {
            entry.Steps = steps;
        }
        else
        {
            entries.Add(new Entry { Date = dateStr, Steps = steps });
        }
        SaveEntries(entries);
        double dist = steps * cfg.StrideCm / 100000.0;
        double cal = steps * 0.04 * cfg.WeightKg / 70.0;
        Console.WriteLine($"✅ Logged {steps:N0} steps on {dateStr} – {dist:F2} km, {cal:F0} kcal");
        return entries;
    }

    static void Stats(List<Entry> entries, Config cfg)
    {
        if (!entries.Any())
        {
            Console.WriteLine("No entries.");
            return;
        }
        int totalSteps = entries.Sum(e => e.Steps);
        double totalDist = totalSteps * cfg.StrideCm / 100000.0;
        double totalCal = totalSteps * 0.04 * cfg.WeightKg / 70.0;
        int avgSteps = totalSteps / entries.Count;
        string today = DateTime.Now.ToString("yyyy-MM-dd");
        var todayEntry = GetEntry(entries, today);
        int todaySteps = todayEntry?.Steps ?? 0;
        int pct = Math.Min(100, (int)((double)todaySteps / cfg.Goal * 100));
        double fuelSaved = totalDist / 10.0;
        double co2Saved = fuelSaved * 2.31;

        Console.WriteLine($"\n🚶 Pedometer (Energy Saving)");
        Console.WriteLine($"Today: {todaySteps:N0} steps");
        Console.WriteLine($"Goal: {cfg.Goal:N0} steps ({pct}% complete)");
        Console.WriteLine($"\n📊 Statistics");
        Console.WriteLine($"Total steps: {totalSteps:N0}");
        Console.WriteLine($"Total distance: {totalDist:F2} km");
        Console.WriteLine($"Total calories: {totalCal:F0} kcal");
        Console.WriteLine($"Average: {avgSteps:N0} steps/day");
        Console.WriteLine($"Energy saved: {fuelSaved:F2} L of fuel");
        Console.WriteLine($"CO₂ saved: {co2Saved:F2} kg");
    }

    static void ListEntries(List<Entry> entries, Config cfg)
    {
        if (!entries.Any())
        {
            Console.WriteLine("No entries.");
            return;
        }
        var sorted = entries.OrderByDescending(e => e.Date);
        Console.WriteLine("\n📋 Daily Log:");
        foreach (var e in sorted)
        {
            double dist = e.Steps * cfg.StrideCm / 100000.0;
            double cal = e.Steps * 0.04 * cfg.WeightKg / 70.0;
            int pct = Math.Min(100, (int)((double)e.Steps / cfg.Goal * 100));
            Console.WriteLine($"{e.Date} | {e.Steps:N0} steps | {dist:F2} km | {cal:F0} kcal | {pct}%");
        }
    }

    static void ExportCSV(List<Entry> entries, Config cfg, string filename)
    {
        using var writer = new StreamWriter(filename);
        writer.WriteLine("Date,Steps,Distance (km),Calories,Goal %");
        foreach (var e in entries)
        {
            double dist = e.Steps * cfg.StrideCm / 100000.0;
            double cal = e.Steps * 0.04 * cfg.WeightKg / 70.0;
            int pct = Math.Min(100, (int)((double)e.Steps / cfg.Goal * 100));
            writer.WriteLine($"{e.Date},{e.Steps},{dist:F2},{cal:F0},{pct}");
        }
        Console.WriteLine($"✅ Exported {entries.Count} entries to {filename}");
    }
}
